### DISCUSSION TITLE
Did grblHAL support rigid Threading or rigid Tapping ?

### DISCUSSION THREAD

**Discussion by hanifhan** [grblHAL/core#295] [General]
Hi terje , id like to ask if is there any function to activate rigid tapping or threading ?

**Comment by terjeio** [comment]
[Some drivers](https://github.com/grblHAL/drivers) supports threading (spindle synced motion) in lathe mode (still needs to be tested extensively). Rigid tapping or threading in a mill is not - but might be supported later.

**Reply by hanifhan** [reply]
oh i seee , and then how about activating spindle sync in mill terje , i ve try to activate it , but it seems there is an eror like this
![image](https://user-images.githubusercontent.com/78627497/235619455-dd892c17-1943-481d-affc-a6fd79959caa.png)

**Reply by terjeio** [reply]
Update to the latest version, including the submodules.

**Reply by hanifhan** [reply]
hi terje , i've updated the submodules , and try to activate it , but it seems after i update the submodules & driver ,( btw im using nucleo stm32f446re  controller ) my controller often get freezing and get no respon  , and is spindle sync support this controller ?

**Reply by terjeio** [reply]
Which board map? Does it freeze when sitting idle or while running gcode?
Yes, spindle sync is supported - I have recently switched to a Nucleo STM32F446 in my lathe, but have not finished testing yet.

**Reply by hanifhan** [reply]
Im using protoneer , its in idle mode , and oftenly get freezing , im use it for milling , how to activate it ? can u give me a clue ?

**Reply by hanifhan** [reply]
im adding the pin spindle ( SPINDLE INDEX PORT & SPINDLE PULSE PORT ) on protoneer_3xx board map

**Reply by terjeio** [reply]
Note that you do not have many options for SPINDLE PULSE PORT/PIN - PA15 or PD2 for now. The signal has to be "routed" as a clock input to a timer.

Freezing is odd, I have not seen that before. Have you excluded EMI (interference) as a source? No floating inputs, proper grounding, good USB cable, ...

**Reply by hanifhan** [reply]
ooh actually my cable was bad , thats why its often got freezing , and  im forgot to settings the spindle ppr ( $38 ) , thats why my respon of encodor not displayed , but i found another problem , under the 600 rpm the read seems accurate, but  after 1000K rpm , the accuracy of the read seems decrease and got until +- 200 rpm , should i set anything to get more accurate counting ?

**Reply by terjeio** [reply]
What is your PPR setting?

**Reply by hanifhan** [reply]
sorry for late reply , im using 2500 PPR encoder

**Reply by hanifhan** [reply]
for example , when my spindle run 1900 its read 2000 on encoder , and when its run 2000 the encoder still read 2000 , and when its run 2100 its read 2000 or 2182 , btw my speed at tolerance is 0

**Reply by terjeio** [reply]
$38 is set to 2500 or 10000?
Are the signals clean above 2000 RPM?
Max RPM rating of the encoder is?

I have to rig my simulator to check if this is a firmware issue, give me some time for that.

**Reply by hanifhan** [reply]
my $38 is set on 2500

**Reply by hanifhan** [reply]
yea , my encoder singnal seems good , hmm im not sure about my max rpm rating , but im sure that my encoder could handle more than 3000 rpm , im only using 3000 max rpm for my spindle

**Reply by hanifhan** [reply]
how is the spindle linearization work terje ? if i sent rpm 100 and my encoder is read 200 , did the encoder control the output so the motor will decrease the speed to 100 automatically or we should sent a code for the synchronization ?

**Reply by terjeio** [reply]
> how is the spindle linearization work 

See [this disussion.](https://github.com/grblHAL/core/discussions/269)

I made an attemt at PID controlled feedback for the MSP432 driver, but have not explored that any further than as an initial experiment. [The code is still there](https://github.com/grblHAL/MSP432P401R/blob/a06ae22d9f9b1d6a52ce02d1b6f94ded16458fbb/driver.c#L1970-L2006) and if someone would like to try porting it to another driver that should not be too hard. Search for `SPINDLE_RPM_CONTROLLED` in  _driver.c_ for the other bits of code.

**Reply by hanifhan** [reply]
im typo , i mean how is the spindle synchronization work ?

**Reply by terjeio** [reply]
Short version: Step generation breaks motion up in ~5ms segments where step rate and direction remains the same. For each new segment angular position is read from the encoder input and compared with the position the stepper driver says it should be. The difference is fed to a [PID controller](https://www.wescottdesign.com/articles/pid/pidWithoutAPhd.pdf)  (pdf) and the result is used to calculate and set a step rate that would lead to the positions (eventually) matching up. Most of the code is in _driver.c_, [PID controller code](https://github.com/grblHAL/core/blob/master/pid.c) is in the core.

Note that the PID controller/feedback loop [has to be tuned](https://duckduckgo.com/?t=ffab&q=PID+tuning&ia=web) to match the dynamic properties of the system. [ioSender](https://github.com/terjeio/ioSender) has a tab to aid tuning where actual and programmed position is plotted, available if [PID logging is enabled](https://github.com/grblHAL/core/blob/af2a922915e4564d812f210184c58e30d07292bc/config.h#L499-L500).

**Reply by hanifhan** [reply]
so , is it automatically synchronizing the spindle speed based on Encoder or the function still not enable ?

**Reply by hanifhan** [reply]
can u give me an example tutorial ?

**Reply by terjeio** [reply]
> so , is it automatically synchronizing the spindle speed based on Encoder...

It is synchronizing the feed rate with the spindle speed - or more precisely, the feed rate to keep the tool tip in sync with the angular position of the spindle.

> ...  or the function still not enable ?

Sorry, I do not understand this part of the question. Or are you asking about closed loop spindle speed where the encoder input is used to keep the spindle at the programmed speed? If so, then this functionality is only available for the MSP432 driver and it is in an experimental state.

> can u give me an example tutorial ?

I intend to write a Wiki page about how to configure and tune spindle sync. Or did you think of something else?

**Reply by hanifhan** [reply]
ohh i see , i think the spindly sync is the function where the encoder input is used to keep the spindle at programmed speed

**Comment by tggo** [comment]
I try spindle sync with Nucleo-144 (STM32F756ZG MCU)  and Omron E6B2-CWZ6C 500p/r (with one output OUT A) and index hall sensor index. And receive odd numbers with constant spindle speed.

All cables fine, tested encoder with ESP32  Pulse Counter (PCNT) - all work fine.


Odd rpm count 
```
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,1208|Pn:S|A:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,360|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,5952|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,151|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,1235|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,1068|Pn:S|WCO:-82.000,0.000,-7.200>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,166667|Pn:S|Ov:100,100,100|A:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,7246|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,288|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,27778|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,7246|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,4902|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,1603|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,1263|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,288|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,1312|Pn:S|WCO:-82.000,0.000,-7.200>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,7246|Pn:S|Ov:100,100,100|A:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,5747|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,5747|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,1111|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,4902|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,158|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,1323|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,380|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,395|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,963|Pn:S|WCO:-82.000,0.000,-7.200>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,307|Pn:S|Ov:100,100,100|A:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,5747|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,5952|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,4902|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,135|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,189|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,1208|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,294|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,1157|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,1029|Pn:S|WCO:-82.000,0.000,-7.200>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,7246|Pn:S|Ov:100,100,100|A:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,9804|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,1149|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,1253|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,1253|Pn:S>
```

my_machine_map.h 
```
#define SPINDLE_INDEX_PORT      GPIOB
#define SPINDLE_INDEX_PIN       15
#define SPINDLE_INDEX_BIT       (1<<SPINDLE_INDEX_PIN)

#define SPINDLE_PULSE_PORT      GPIOD
#define SPINDLE_PULSE_PIN       2
#define SPINDLE_PULSE_BIT       (1<<SPINDLE_PULSE_PIN)
```

Can you please tell me what I need to look at and configure in order for it to work?

**Reply by tggo** [reply]
Oh... 
I was add resistor from +V to Out A :) start work more realistic.

But real RPM is 214
```
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,236|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,214|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,295|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,240|Pn:ZS|WCO:-82.000,0.000,-7.200>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,229|Pn:ZS|Ov:100,100,100|A:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,219|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,215|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,264|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,237|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,216|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,296|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,216|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,1104|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,235|Pn:ZS|WCO:-82.000,0.000,-7.200>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,265|Pn:ZS|Ov:100,100,100|A:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,218|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,242|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,266|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,265|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,211|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,335|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,294|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,390|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,217|Pn:ZS|WCO:-82.000,0.000,-7.200>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,344|Pn:ZS|Ov:100,100,100|A:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,805|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,235|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,233|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,262|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,265|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,337|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,233|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,296|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,239|Pn:ZS|WCO:-82.000,0.000,-7.200>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,224|Pn:ZS|Ov:100,100,100|A:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,402|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,346|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,401|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,337|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,292|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,214|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,200,383|Pn:ZS>
```

**Reply by tggo** [reply]
Yep, after changing the connection point of the encoder cable shield, everything became great.

Real RPM 110
```
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,107|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,106|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,109|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,104|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,107|Pn:S|WCO:-82.000,0.000,-7.200>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,107|Pn:S|Ov:100,100,100|A:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,108|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,109|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,109|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,105|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,105|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,107|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,109|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,110|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,105|Pn:S|WCO:-82.000,0.000,-7.200>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,109|Pn:S|Ov:100,100,100|A:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,108|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,107|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,108|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,107|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,109|Pn:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,100,108|Pn:S>
```

Thanks for the help!

**Reply by tggo** [reply]
And another problem, receive - **Status_GcodeSpindleNotRunning = 41**
Why? Spindle work normal.  RPM is real.

```
M3 S100
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,110,106|Pn:ZS|Ov:100,100,100|A:S>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,110,107|Pn:ZS>
<Idle|MPos:0.000,0.000,0.000|Bf:35,1023|FS:0,110,109|Pn:ZS>
G33 K1 Z-10
error:41
```


```
>>> $sd
[SPINDLE:166,16672,0,166.727]
ok
>>> $sd
[SPINDLE:182,18246,0,182.467]
ok
>>> $sd
[SPINDLE:195,19555,0,195.554]
ok
>>> $sd
[SPINDLE:208,20864,0,208.647]
ok
>>> $sd
[SPINDLE:216,21669,0,216.690]
```

```$0 = 10.0    (Step pulse time, microseconds)
$1 = 255    (Step idle delay, milliseconds)
$2 = 0    (Step pulse invert, mask)
$3 = 0    (Step direction invert, mask)
$4 = 7    (Invert step enable pin, boolean)
$5 = 0    (Invert limit pins, boolean)
$9 = 1   
$10 = 511    (Status report options, mask)
$11 = 0.010    (Junction deviation, millimeters)
$12 = 0.002    (Arc tolerance, millimeters)
$13 = 0    (Report in inches, boolean)
$14 = 0   
$15 = 0   
$16 = 0   
$17 = 0   
$20 = 0    (Soft limits enable, boolean)
$18 = 0   
$21 = 0    (Hard limits enable, boolean)
$22 = 0    (Homing cycle enable, boolean)
$23 = 0    (Homing direction invert, mask)
$24 = 25.0    (Homing locate feed rate, mm/min)
$25 = 500.0    (Homing search seek rate, mm/min)
$26 = 250    (Homing switch debounce delay, milliseconds)
$27 = 1.000    (Homing switch pull-off distance, millimeters)
$28 = 0.100   
$29 = 0.0   
$30 = 760.000    (Maximum spindle speed, RPM)
$31 = 80.000    (Minimum spindle speed, RPM)
$32 = 2    (Laser-mode enable, boolean)
$33 = 10000.0   
$34 = 0.0   
$35 = 11.0   
$36 = 100.0   
$38 = 360   
$37 = 0   
$39 = 1   
$40 = 0   
$43 = 1   
$44 = 4   
$45 = 3   
$46 = 0   
$62 = 0   
$63 = 2   
$64 = 0   
$70 = 3   
$90 = 0.500   
$91 = 0.500   
$92 = 0.500   
$95 = 0.000   
$100 = 3200.000    (X-axis travel resolution, step/mm)
$101 = 3200.000    (Y-axis travel resolution, step/mm)
$102 = 3200.000    (Z-axis travel resolution, step/mm)
$110 = 1000.000    (X-axis maximum rate, mm/min)
$111 = 500.000    (Y-axis maximum rate, mm/min)
$112 = 1000.000    (Z-axis maximum rate, mm/min)
$120 = 10.000    (X-axis acceleration, mm/sec^2)
$121 = 10.000    (Y-axis acceleration, mm/sec^2)
$122 = 100.000    (Z-axis acceleration, mm/sec^2)
$130 = 100.000    (X-axis maximum travel, millimeters)
$131 = 200.000    (Y-axis maximum travel, millimeters)
$132 = 300.000    (Z-axis maximum travel, millimeters)
$300 = grblHAL   
$301 = 0   
$302 = 192.168.88.98   
$303 = 192.168.88.1   
$304 = 255.255.255.0   
$305 = 23   
$307 = 80   
$340 = 0.0   
$341 = 0   
$342 = 30.0   
$343 = 25.0   
$344 = 200.0   
$345 = 200.0   
$346 = 1   
$384 = 0   
$398 = 35   
$481 = 0 
```

**Reply by tggo** [reply]
Record video with [debug session](https://youtu.be/jh0yEYxLT_A) https://youtu.be/jh0yEYxLT_A

**Reply by terjeio** [reply]
Do you have a spindle spin up delay after starting the spindle (G4Pn where n is number of seconds) or set $340 (spindle at speed tolerance)?
Setting $340 > 0 is the safest option as this checks the actual RPM vs the programmed RPM before continuing.

**Reply by tggo** [reply]
Ok,  $340=50 help
Thanks so much.


Now freezing  after some work, may be wire not good. I will check.
Thanks again.
